### PR TITLE
chore(getConfiguration): burn it to the ground 🔥

### PR DESCRIPTION
**Pull request by Haroenv**
1. remove `getConfiguration` in every widget
2. remove `getConfiguration` from types & expectations
3. change tests which were using `getConfiguration` to `getWidgetSearchParameters`

### CODE REVIEW

File: src/widgets/range-slider/__tests__/range-slider-test.js [hunk 1]
```diff
@@ -62,15 +62,20 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/range-slide
     });
 
     describe('min option', () => {
-      it('refines when no previous configuration', () => {
+      // @TODO: gWSP does not yet take min & max in account
+      it.skip('refines when no previous configuration', () => {
```
Comment (Haroenv) [inline src/widgets/range-slider/__tests__/range-slider-test.js]: these need to be fixed in a separate PR, cc @samouss

File: src/connectors/range/connectRange.js [hunk 2]
```diff
@@ -275,9 +241,17 @@ export default function connectRange(renderFn, unmountFn = noop) {
...
 
-        return state
-          .removeNumericRefinement(attribute)
-          .removeDisjunctiveFacet(attribute);
+        const stateWithoutDisjunctive = state.removeDisjunctiveFacet(attribute);
+
+        // can not use setQueryParameters || removeNumericRefinements, because
+        // they both keep the old value. This isn't immutable, but it is fine
+        // since it's already a copy.
+        stateWithoutDisjunctive.numericRefinements = {
```
Comment (Haroenv) [inline src/connectors/range/connectRange.js]: I'm not sure if we can find a better solution for this

File: src/widgets/range-slider/__tests__/range-slider-test.js [hunk 3]
```diff
@@ -268,9 +297,17 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/range-slide
...
       it('configures the disjunctiveFacets', () => {
-        expect(widget.getConfiguration(new SearchParameters())).toEqual(
+        expect(
+          widget.getWidgetSearchParameters(new SearchParameters(), {
+            uiState: {},
+          })
+        ).toEqual(
           new SearchParameters({
             disjunctiveFacets: ['aNumAttr'],
+            // @TODO: gWSP does not yet take min & max in account (so is this correct?)
```
Comment (Haroenv) [inline src/widgets/range-slider/__tests__/range-slider-test.js]: needs to be fixed in a separate PR first

File: src/types/widget.ts [hunk 4]
```diff
@@ -129,11 +129,24 @@ export interface Widget {
...
+   * the search parameters applied.
+   * @param uiState current state
+   * @param widgetStateOptions extra information to calculate uiState
+   */
   getWidgetState?(
     uiState: IndexUiState,
     widgetStateOptions: WidgetStateOptions
   ): IndexUiState;
+  /**
+   * This function is required for a widget to behave correctly when a URL is
```
Comment (samouss) [inline src/types/widget.ts]: This is half true since we know use this function to sets the configuration too.

File: src/connectors/menu/__tests__/connectMenu-test.js [hunk 5]
```diff
@@ -25,7 +25,7 @@ describe('connectMenu', () => {
   //   const helper = jsHelper(
   //     {},
   //     '',
-  //     widget.getConfiguration(new SearchParameters())
+  //     widget.getWidgetSearchParameters(new SearchParameters(),{uiState:{}})
```
Comment (samouss) [inline src/connectors/menu/__tests__/connectMenu-test.js]: Either use or remove the commented function.

File: src/connectors/hierarchical-menu/__tests__/connectHierarchicalMenu-test.js [hunk 6]
```diff
@@ -18,7 +18,7 @@ describe('connectHierarchicalMenu', () => {
   //   const helper = jsHelper(
   //     {},
   //     '',
-  //     widget.getConfiguration(new SearchParameters())
+  //     widget.getWidgetSearchParameters(new SearchParameters(),{uiState:{}})
```
Comment (samouss) [inline src/connectors/hierarchical-menu/__tests__/connectHierarchicalMenu-test.js]: Either use or remove the commented function.

File: src/connectors/range/connectRange.js (hunk 2 above)
Comment (samouss) [reply src/connectors/range/connectRange.js]: IIRC you can use `removeNumericRefinements` with the attribute only, it will perform a clear under the hood.

File: src/types/widget.ts (hunk 4 above)
Comment (Haroenv) [reply src/types/widget.ts]: Do you have a better description? “Also used to keep track (or merge) of refinements when other widgets are added/removed”?

File: src/connectors/range/connectRange.js (hunk 2 above)
Comment (Haroenv) [reply src/connectors/range/connectRange.js]: It doesn’t actually remove the refinement, I havent found a solution which doesn’t break for refining to an empty state

File: src/widgets/range-slider/__tests__/range-slider-test.js (hunk 1 above)
Comment (Haroenv) [reply src/widgets/range-slider/__tests__/range-slider-test.js]: https://github.com/algolia/instantsearch.js/pull/4088

File: src/widgets/range-slider/__tests__/range-slider-test.js (hunk 3 above)
Comment (Haroenv) [reply src/widgets/range-slider/__tests__/range-slider-test.js]: https://github.com/algolia/instantsearch.js/pull/4088

File: src/connectors/range/connectRange.js [hunk 7]
```diff
@@ -275,9 +241,17 @@ export default function connectRange(renderFn, unmountFn = noop) {
       dispose({ state }) {
         unmountFn();
 
-        return state
-          .removeNumericRefinement(attribute)
-          .removeDisjunctiveFacet(attribute);
+        const stateWithoutDisjunctive = state.removeDisjunctiveFacet(attribute);
+
+        // can not use setQueryParameters || removeNumericRefinements, because
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/connectors/range/connectRange.js]: thanks!

File: src/connectors/breadcrumb/__tests__/connectBreadcrumb-test.js [hunk 8]
```diff
@@ -1018,67 +789,40 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/breadcrumb/
...
-        hierarchicalFacets: [
-          {
-            attributes: ['otherCategory', 'otherSub_category'],
-            name: 'otherCategory',
-            separator: ' > ',
-          },
-        ],
-      })
-    );
+    it('does not remove refinement', () => {
```
Comment (tkrugg) [inline src/connectors/breadcrumb/__tests__/connectBreadcrumb-test.js]: not sure I get this part.
Removing the widget should remove the refinement, no?

File: src/connectors/hits-per-page/__tests__/connectHitsPerPage-test.js [hunk 9]
```diff
@@ -234,10 +242,13 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/hits-per-pa
...
     expect(
-      widget.getConfiguration(new SearchParameters({ hitsPerPage: 20 }))
+      widget.getWidgetSearchParameters(
+        new SearchParameters({ hitsPerPage: 20 }),
+        { uiState: {} }
+      )
     ).toEqual(
       new SearchParameters({
-        hitsPerPage: 20,
+        hitsPerPage: 10,
```
Comment (tkrugg) [inline src/connectors/hits-per-page/__tests__/connectHitsPerPage-test.js]: isn't this supposed to be 20?
this doesn't look like it overrides it, right?

File: src/connectors/breadcrumb/__tests__/connectBreadcrumb-test.js (hunk 8 above)
Comment (Haroenv) [reply src/connectors/breadcrumb/__tests__/connectBreadcrumb-test.js]: no, a breadcrumb's refinement is set by a configure or hierarchical facet, and should thus stay. This didn't change in this PR by the way, it was simply not tested

File: src/connectors/hits-per-page/__tests__/connectHitsPerPage-test.js (hunk 9 above)
Comment (Haroenv) [reply src/connectors/hits-per-page/__tests__/connectHitsPerPage-test.js]: we override the previously set hPP with the "default" one, this was defined by #4069
